### PR TITLE
Verify transient runtime module dependencies

### PR DESCRIPTION
**Pull request by ericmj**

### CODE REVIEW

File: lib/mix/lib/mix/compilers/elixir.ex [hunk 1]
```diff
@@ -236,29 +236,39 @@ defmodule Mix.Compilers.Elixir do
...
+
+  defp fixpoint_runtime_modules(sources, changed, dependent, not_dependent) do
+    {new_dependent, not_dependent} =
+      Enum.reduce(not_dependent, {dependent, []}, fn module, {new_dependent, not_dependent} ->
+        depending? =
+          Enum.any?(module(module, :sources), fn file ->
+            source(runtime_references: runtime_refs) =
+              List.keyfind(sources, file, source(:source))
+
+            Enum.any?(changed, fn {module(module: module), true} -> module in runtime_refs end)
```
Comment (josevalim) [inline lib/mix/lib/mix/compilers/elixir.ex]: has_any_key doesn't work here?

Comment (ericmj) [reply lib/mix/lib/mix/compilers/elixir.ex]: No, the keys are `module()` records and we have a list of module atoms.

Comment (josevalim) [reply lib/mix/lib/mix/compilers/elixir.ex]: Oh, I see. I think we can still store only the module name in the dependent map, no?

Comment (josevalim) [reply lib/mix/lib/mix/compilers/elixir.ex]: Something like this (untested):

```elixir
  defp dependent_runtime_modules(sources, all_modules, pending_modules) do
    changed_modules =
      for module(module: module) = entry <- all_modules,
          entry not in pending_modules,
          into: %{},
          do: {module, true}

    fixpoint_runtime_modules(sources, changed_modules, %{}, pending_modules)
  end

  defp fixpoint_runtime_modules(sources, changed, dependent, not_dependent) do
    {new_dependent, not_dependent} =
      Enum.reduce(not_dependent, {dependent, []}, fn module, {new_dependent, not_dependent} ->
        depending? =
          Enum.any?(module(module, :source), fn file ->
            source(runtime_references: runtime_refs) =
              List.keyfind(sources, file, source(:source))

            has_any_key(changed, runtime_refs)
          end)

        if depending? do
          {Map.put(new_dependent, module(module, :module), true), not_dependent}
        else
          {new_dependent, [module | not_dependent]}
        end
      end)

    if map_size(dependent) != map_size(new_dependent) do
      fixpoint_runtime_modules(sources, new_dependent, new_dependent, not_dependent)
    else
      Map.keys(new_dependent)
    end
  end
```

**Pull request merged**